### PR TITLE
add b3 propagator for tracing

### LINKED ISSUE #725: Support B3 headers in OTEL propagation
**Problem**

Tracing is instrumental in supported distributed systems such as Knative Eventing, Kafka, etc.   Many systems (including Istio), by default use B3 (multi) headers.   Kafka Broker appears to be configured for only W3C headers.   Traces involving Knative Eventing stop when it hits the broker or only start after messages hit other systems, you don't get an end-to-end trace when using B3

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?
* Producer
* Consumer

**Exit Criteria**
Kafka Broker propagates B3 headers (in addition to W3C)

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by snowwolf007cn**
Fixes #725 

## Proposed Changes

- add heads-format config entry in config-tracing, default is 'w3c'
- when the backend is 'zipkin', 'b3-multi-header' and 'b3-single-header' is supported to propagate with b3 headers
- config propagator to use 'w3c' or 'b3'

**Release Note**

```release-note
🎁 you can now config which header format(b3 multi header, b3 single header, and w3c trace-context) to be used while using Zipkin backend
```

**Docs**

### CODE REVIEW

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java [hunk 1]
```diff
@@ -122,9 +127,14 @@ public OpenTelemetrySdk setup() {
...
-    sdkBuilder.setPropagators(ContextPropagators.create(
-      W3CTraceContextPropagator.getInstance()
-    ));
+
+    ContextPropagators contextPropagators;
+    switch (getHeadsFormat()) {
+      case B3_MULTI_HEADER -> contextPropagators = ContextPropagators.create(B3Propagator.injectingMultiHeaders());
+      case B3_SINGLE_HEADER -> contextPropagators = ContextPropagators.create(B3Propagator.injectingSingleHeader());
+      default -> contextPropagators = ContextPropagators.create(W3CTraceContextPropagator.getInstance());
+    }
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java:136]: This looks a little bit easier to read:
```suggestion
    final var contextPropagators = switch (getHeadsFormat()) {
      case B3_MULTI_HEADER -> ContextPropagators.create(B3Propagator.injectingMultiHeaders());
      case B3_SINGLE_HEADER -> ContextPropagators.create(B3Propagator.injectingSingleHeader());
      default -> ContextPropagators.create(W3CTraceContextPropagator.getInstance());
    };
```

File: data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java [hunk 2]
```diff
@@ -39,12 +39,14 @@ public void shouldParseConfigGivenDirectory() throws IOException {
     write(dir, "/backend", " zipkin ");
     write(dir, "/sample-rate", " 0.1 ");
     write(dir, "/zipkin-endpoint", "  http://localhost:9241/v2/api/spans/     ");
+    write(dir, "/heads-format", " b3-multi-header  ");
```
Comment (pierDipi) [inline data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java:42]: We should test also w3 (default) and b3 single header, we can keep this test as it was and assert that `HeadersFormat` is `W3C` and add these 2 more tests:
```java
  @Test
  public void shouldParseConfigB3SingleHeaderGivenDirectory() throws IOException {

    final var dir = Files.createTempDirectory("tracing");

    write(dir, "/backend", " zipkin ");
    write(dir, "/sample-rate", " 0.1 ");
    write(dir, "/zipkin-endpoint", "  http://localhost:9241/v2/api/spans/     ");
    write(dir, "/headers-format", " b3-single-header  ");

    final var config = TracingConfig.fromDir(dir.toAbsolutePath().toString());

    assertThat(config.getBackend()).isEqualTo(Backend.ZIPKIN);
    assertThat(config.getSamplingRate()).isEqualTo(0.1F);
    assertThat(config.getUrl()).isEqualTo("http://localhost:9241/v2/api/spans/");
    assertThat(config.getHeadsFormat()).isEqualTo(HeadersFormat.B3_SINGLE_HEADER);
  }

  @Test
  public void shouldParseConfigB3MultiHeaderGivenDirectory() throws IOException {

    final var dir = Files.createTempDirectory("tracing");

    write(dir, "/backend", " zipkin ");
    write(dir, "/sample-rate", " 0.1 ");
    write(dir, "/zipkin-endpoint", "  http://localhost:9241/v2/api/spans/     ");
    write(dir, "/headers-format", " b3-multi-header  ");

    final var config = TracingConfig.fromDir(dir.toAbsolutePath().toString());

    assertThat(config.getBackend()).isEqualTo(Backend.ZIPKIN);
    assertThat(config.getSamplingRate()).isEqualTo(0.1F);
    assertThat(config.getUrl()).isEqualTo("http://localhost:9241/v2/api/spans/");
    assertThat(config.getHeadsFormat()).isEqualTo(HeadersFormat.B3_MULTI_HEADER);
  }
```

**Review by pierDipi — COMMENTED** [review]
Awesome @snowwolf007cn !

I left some comments inline and I'd also rename everywhere `headsFormat` to `headersFormat` for better clarity.

**Comment by pierDipi** [comment]
@snowwolf007cn there is a problem with the CLA, please take a look: https://github.com/knative-sandbox/eventing-kafka-broker/pull/1546/checks?check_run_id=4331388121

File: data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java (hunk 2 above)
Comment (snowwolf007cn) [reply data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java]: reasonable, I've updated the PR.

**Comment by snowwolf007cn** [comment]
> @snowwolf007cn there is a problem with the CLA, please take a look: https://github.com/knative-sandbox/eventing-kafka-broker/pull/1546/checks?check_run_id=4331388121

I think you've rescanned.

File: data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java [hunk 3]
```diff
@@ -71,6 +133,14 @@ public void shouldReturnUnknownBackendWhenNoFilesArePresent() throws IOException
     assertThat(config.getBackend()).isEqualTo(Backend.UNKNOWN);
   }
 
+  @Test
+  public void shouldReturnW3CHeadsFormatWhenDirectoryIsNotPresent() throws IOException {
```
Comment (pierDipi) [inline data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java:137]:
```suggestion
  public void shouldReturnW3CHeadersFormatWhenDirectoryIsNotPresent() throws IOException {
```

File: data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java [hunk 4]
```diff
@@ -55,12 +93,36 @@ public void shouldReturnUnknownBackendWhenBackendContainsUnknownName() throws IO
...
+    write(dir, "/zipkin-endpoint", "  http://localhost:9241/v2/api/spans/     ");
+    write(dir, "/headers-format", " b3-multi-header ");
+
+    final var config = TracingConfig.fromDir(dir.toAbsolutePath().toString());
+
+    assertThat(config.getHeadersFormat()).isEqualTo(HeadersFormat.W3C);
+  }
+
+  @Test
+  public void shouldReturnW3CHeadsFormatWhenNoFilesArePresent() throws IOException {
```
Comment (pierDipi) [inline data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java:118]:
```suggestion
  public void shouldReturnW3CHeadersFormatWhenNoFilesArePresent() throws IOException {
```

File: data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java [hunk 5]
```diff
@@ -55,12 +93,36 @@ public void shouldReturnUnknownBackendWhenBackendContainsUnknownName() throws IO
...
     write(dir, "/zipkin-endpoint", "  http://localhost:9241/v2/api/spans/     ");
+    write(dir, "/headers-format", " b3-multi-header ");
 
     final var config = TracingConfig.fromDir(dir.toAbsolutePath().toString());
 
     assertThat(config.getBackend()).isEqualTo(Backend.UNKNOWN);
   }
 
+  @Test
+  public void shouldReturnW3CHeadsFormatWhenBackendContainsUnknownName() throws IOException {
```
Comment (pierDipi) [inline data-plane/core/src/test/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfigTest.java:104]:
```suggestion
  public void shouldReturnW3CHeadersFormatWhenBackendContainsUnknownName() throws IOException {
```

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java [hunk 6]
```diff
@@ -237,14 +260,25 @@ public static TracingConfig fromDir(final String path) throws IOException {
 
     if (backend.equals(Backend.ZIPKIN)) {
       final var zipkinPath = pathOf(path, "zipkin-endpoint");
+      final var headsFormatPath = headersFormatPath(path);
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java:263]: `heads` -> `headers`

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java [hunk 7]
```diff
@@ -237,14 +260,25 @@ public static TracingConfig fromDir(final String path) throws IOException {
...
       if (Files.exists(zipkinPath)) {
         try (final var url = new FileInputStream(zipkinPath.toString())) {
           endpoint = Parser.URL(url);
         }
       }
+      if (Files.exists(headsFormatPath)) {
+        try (final var headsFormatFile = new FileInputStream(headsFormatPath.toString())) {
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java:270]: `heads` -> `headers`

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java [hunk 8]
```diff
@@ -213,19 +235,20 @@ private static String trim(InputStream in) throws IOException {
...
     final var backendPath = backendPath(path);
     if (!Files.exists(backendPath)) {
-      return new TracingConfig(Backend.UNKNOWN, null, 0);
+      return new TracingConfig(Backend.UNKNOWN, null, 0, HeadersFormat.W3C);
     }
 
     var sampleRate = 0F;
     var backend = Backend.UNKNOWN;
     var endpoint = "";
+    var headsFormat = HeadersFormat.W3C;
```
Comment (pierDipi) [inline data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java:244]: `heads` -> `headers`

**Review by pierDipi — COMMENTED** [review]
/ok-to-test

Some more renaming.

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java (hunk 8 above)
Comment (snowwolf007cn) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/tracing/TracingConfig.java]: solved.

**Comment by pierDipi** [comment]
@snowwolf007cn the failure wants you to regenerate the file with third-party licenses by running this command:
```
./mvnw -Dlicense.outputDirectory=. license:aggregate-add-third-party
```

**Comment by pierDipi** [comment]
inside the `data-plane` directory.

**Review by pierDipi — APPROVED** [review]
/lgtm
/approve

Thank you!

**Comment by hassanbenhzaine** [comment]
Hello everyone, could someone please show me how to replace knative propagator to use B3 instead if W3C by default

**Pull request merged**